### PR TITLE
Add image compression module

### LINKED ISSUE #3: Add support for compressing the photo which exceed upload limit
[Issuehunt badges](https://issuehunt.io/r/LitoMore/fanfou-pro/issues/3)

The size limit of the static image by the Fanfou API is: `3 MB`

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $20.00)

- [litomore litomore](https://issuehunt.io/u/litomore) ($20.00)

### Submitted pull Requests
- [#4 Add image compression module](https://issuehunt.io/r/LitoMore/fanfou-pro/pull/4)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/LitoMore/fanfou-pro/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by twoheartliu**
Resolve #3

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#3: Add support for compressing the photo which exceed upload limit](https://issuehunt.io/repos/244042026/issues/3)
---
</details>

### CODE REVIEW

File: src/components/post-form.js [hunk 1]
```diff
@@ -5,7 +5,7 @@ import styled from 'styled-components';
 import {LoadingOutlined} from '@ant-design/icons';
 import slogan from '../assets/slogan.svg';
 import uploadIcon from '../assets/upload-icon.svg';
-
+import {fileToBase64ByQuality, convertBase64UrlToBlob} from '../utils/image-compression';
```
Comment (LitoMore) [inline src/components/post-form.js:8]:
```suggestion
import {fileToBase64ByQuality, convertBase64UrlToBlob} from '../utils/image-compression';

```

Comment (LitoMore) [inline src/components/post-form.js:8]: An empty line required.

```suggestion
import {fileToBase64ByQuality, convertBase64UrlToBlob} from '../utils/image-compression';

```

File: src/components/post-form.js [hunk 2]
```diff
@@ -64,8 +64,24 @@ class PostForm extends React.Component {
 	handleUpload = event => {
 		const {files} = event.target;
 		const {setFile} = this.props;
+
+		console.log(files[0]);
```
Comment (LitoMore) [inline src/components/post-form.js]: Remove this `console.log`.

```suggestion
```

File: src/utils/image-compression.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+			});
+			fileReader.addEventListener('error', event => {
+				reject(event);
+			});
+			fileReader.readAsDataURL(file);
+		}
+	});
+};
+
+//  图片最大宽度
```
Comment (LitoMore) [inline src/utils/image-compression.js]: Do not leave any Chinese comment.

File: src/utils/image-compression.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+const MAX_WIDTH = 1200;
+
+export const compress = (base64, quality, mimeType) => {
+	console.log('mimeType', mimeType);
+	const cvs = document.createElement('canvas');
+	const img = document.createElement('img');
+	img.crossOrigin = 'anonymous';
+	return new Promise(resolve => {
+		img.src = base64;
+		// 图片偏移值
```
Comment (LitoMore) [inline src/utils/image-compression.js]: Ditto. Use English.

File: src/utils/image-compression.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+		// const offetX = 0;
```
Comment (LitoMore) [inline src/utils/image-compression.js]: Remove this unused code.

File: src/utils/image-compression.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+		img.addEventListener('load', () => {
+			if (img.width > MAX_WIDTH) {
+				cvs.width = MAX_WIDTH;
+				cvs.height = img.height * MAX_WIDTH / img.width;
+				// Const offetX = (img.width - MAX_WIDTH) / 2;
```
Comment (LitoMore) [inline src/utils/image-compression.js]: Ditto.

File: src/utils/image-compression.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+			fileReader.readAsDataURL(file);
+		}
+	});
+};
+
+//  图片最大宽度
+const MAX_WIDTH = 1200;
+
+export const compress = (base64, quality, mimeType) => {
+	console.log('mimeType', mimeType);
```
Comment (LitoMore) [inline src/utils/image-compression.js]: Remove this.
```suggestion
```

File: src/components/post-form.js [hunk 8]
```diff
@@ -64,8 +64,24 @@ class PostForm extends React.Component {
...
 		if (files[0]) {
-			setFile(files[0]);
+			if (files[0].size >= 3000000 && files[0].type !== 'image/gif') {
+				// eslint-disable-next-line
+				const answer = confirm('图片过大，将尝试对图片进行压缩');
+				if (answer === true) {
+					fileToBase64ByQuality(files[0], 100).then(response => {
+						const result_blob = convertBase64UrlToBlob(response, files[0].type);
+						console.log(result_blob);
```
Comment (LitoMore) [inline src/components/post-form.js]: Remove this.

```suggestion
```

**Review by LitoMore — CHANGES_REQUESTED** [review]

File: src/utils/image-compression.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+export const convertBase64UrlToBlob = (base64, mimeType) => {
+	const bytes = window.atob(base64.split(',')[1]);
+	const ab = new ArrayBuffer(bytes.length);
+	const ia = new Uint8Array(ab);
+	for (let i = 0; i < bytes.length; i++) {
+		ia[i] = bytes.charCodeAt(i);
+	}
+
+	const _blob = new Blob([ab], {type: mimeType});
+	console.log(_blob);
```
Comment (LitoMore) [inline src/utils/image-compression.js]: Do not commit any `console.log`.

File: src/components/post-form.js [hunk 10]
```diff
@@ -64,8 +65,23 @@ class PostForm extends React.Component {
...
+					});
+				} else {
+					setFile(null);
+				}
+			} else {
+				setFile(files[0]);
+			}
 		} else {
 			setFile(null);
 		}
```
Comment (LitoMore) [inline src/components/post-form.js:88]: Too much if-else, please optimize your code.

**Review by LitoMore — CHANGES_REQUESTED** [review]

**Comment by twoheartliu** [comment]
Fixed above problem and add jump to image url

File: src/utils/image-compression.js [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+			});
+			fileReader.addEventListener('error', event => {
+				reject(event);
+			});
+			fileReader.readAsDataURL(file);
+		}
+	});
+};
+
+// Setting compression max width
```
Comment (LitoMore) [inline src/utils/image-compression.js]: Remove  this comment

File: src/utils/image-compression.js [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+			cvs.getContext('2d').drawImage(img, 0, 0, cvs.width, cvs.height);
+			const imageData = cvs.toDataURL(mimeType, quality / 100);
+			resolve(imageData);
+		});
+	});
+};
+
+export const convertBase64UrlToBlob = (base64, mimeType) => {
+	const bytes = window.atob(base64.split(',')[1]);
```
Comment (LitoMore) [inline src/utils/image-compression.js]: Danger usage.

File: src/utils/image-compression.js [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+	const ab = new ArrayBuffer(bytes.length);
+	const ia = new Uint8Array(ab);
+	for (let i = 0; i < bytes.length; i++) {
+		ia[i] = bytes.charCodeAt(i);
+	}
+
+	const _blob = new Blob([ab], {type: mimeType});
+	return _blob;
```
Comment (LitoMore) [inline src/utils/image-compression.js]: Unnecessary underscore.

**Review by LitoMore — CHANGES_REQUESTED** [review]

File: src/utils/image-compression.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+			});
+			fileReader.addEventListener('error', event => {
+				reject(event);
+			});
+			fileReader.readAsDataURL(file);
+		}
+	});
+};
+
+export const compress = (base64, quality, mimeType, MAX_WIDTH) => {
```
Comment (LitoMore) [inline src/utils/image-compression.js:19]: maxWidth

**Review by LitoMore — CHANGES_REQUESTED** [review]

**Review by LitoMore — APPROVED** [review]

**Pull request merged**